server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -277,7 +280,21 @@ protected static void parseProperties(ObjectMapper.Builder objBuilder, Map<Strin
                     if (typeParser == null) {
                         throw new MapperParsingException("No handler for type [" + type + "] declared on field [" + fieldName + "]");
                     }
-                    String[] fieldNameParts = fieldName.split("\\.");
+
+                    // We pass explicitly pass the 'limit' parameter to String#split to make sure that the last component is always
+                    // returned, even when it's an empty string. This allows us to handle trailing dots in field names correctly.
+                    String[] fieldNameParts = fieldName.split("\\.", -1);
```

Comment:
Would it be possible to detect empty string parts here already and issue a more comprehensive error message (maybe similar to the IAE that the FieldMapper throws now, but containing the full `fieldName` to make finding the offending field easier e.g. in a large mapping?